### PR TITLE
Use bit mask in Regex's IsWordChar

### PR DESCRIPTION
**Pull request by stephentoub**
Improves throughput of IsWordChar on non-ASCII chars by ~15%.

Roslyn has also fixed the codegen for is/is not patterns, so the switch expressions used here instead have been switched back.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs [hunk 1]
```diff
@@ -1158,20 +1161,26 @@ public static bool IsWordChar(char ch)
...
+                                     ((ulong)UnicodeCategory.DecimalDigitNumber << 8) | ((ulong)UnicodeCategory.ConnectorPunctuation)).AsByte());
+            }
+            else
+#endif
+            {
+                return category is
+                    UnicodeCategory.UppercaseLetter or UnicodeCategory.LowercaseLetter or
+                    UnicodeCategory.TitlecaseLetter or UnicodeCategory.ModifierLetter or
+                    UnicodeCategory.OtherLetter or UnicodeCategory.NonSpacingMark or
+                    UnicodeCategory.DecimalDigitNumber or UnicodeCategory.ConnectorPunctuation;
```
Comment (MihaZupan) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:1183]:
```suggestion
            const int Mask =
                1 << (int)UnicodeCategory.UppercaseLetter |
                1 << (int)UnicodeCategory.LowercaseLetter |
                1 << (int)UnicodeCategory.TitlecaseLetter |
                1 << (int)UnicodeCategory.ModifierLetter |
                1 << (int)UnicodeCategory.OtherLetter |
                1 << (int)UnicodeCategory.NonSpacingMark |
                1 << (int)UnicodeCategory.DecimalDigitNumber |
                1 << (int)UnicodeCategory.ConnectorPunctuation;
            
            return (Mask & (1 << (int)category)) != 0;
```

Is using the vector faster than a simple mask?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:1183]: Hah! Probably not.  Let me check.

Comment (EgorBo) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:1183]: haha, I was about to type the same 😆:
```csharp
static bool Test(char ch, UnicodeCategory category)
{
    if (category <= UnicodeCategory.ConnectorPunctuation)
    {
        uint mask =
            (1 << (int)UnicodeCategory.LowercaseLetter) |
            (1 << (int)UnicodeCategory.TitlecaseLetter) |
            (1 << (int)UnicodeCategory.ModifierLetter) |
            (1 << (int)UnicodeCategory.OtherLetter) |
            (1 << (int)UnicodeCategory.NonSpacingMark) |
            (1 << (int)UnicodeCategory.DecimalDigitNumber) |
            (1 << (int)UnicodeCategory.ConnectorPunctuation);
        return ((mask >> ch) & 1) == 1;
    }
    return false;
}
```
```asm
; Method Prog:Test(ushort,int):bool
G_M42129_IG01:              ;; offset=0000H
G_M42129_IG02:              ;; offset=0000H
       83FA12               cmp      edx, 18
       7F11                 jg       SHORT G_M42129_IG05
G_M42129_IG03:              ;; offset=0005H
       0FB7C1               movzx    rax, cx
       BA3E010400           mov      edx, 0x4013E
       C4E2FBF7C2           shrx     rax, rdx, rax
       83E001               and      eax, 1
G_M42129_IG04:              ;; offset=0015H
       C3                   ret      
G_M42129_IG05:              ;; offset=0016H
       33C0                 xor      eax, eax

G_M42129_IG06:              ;; offset=0018H
       C3                   ret      
; Total bytes of code: 25
```
Technically, JIT emits something similiar for SWITCH opcodes, but Roslyn doesn't want to emit a switch here

Comment (EgorBo) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:1183]: ah, yeah, condition `if (category <= UnicodeCategory.ConnectorPunctuation)` is not needed here.

Still, the codegen can be better if we use `bt` instruction here

**Review by MihaZupan — APPROVED** [review]

**Review by IDisposable — COMMENTED** [review]
There are three blocks of identical code (not counting the source emitting version nor the test) for the WordCategories constant. Could we pull this out somewhere common so we don't eventually get them out of sync?

**Comment by IDisposable** [comment]
Working on the coalescing of the constant... and I noticed a typo.

**Comment by IDisposable** [comment]
@stephentoub is there a performance history that explains the duplicate code in `IsWordChar(char ch)` and `IsBoundaryWordChar(char ch)`? Specifically, is there any reason we _should not_ just make `IsBoundaryWordChar` call into `IsWordChar` like
```csharp
  public static bool IsBoundaryWordChar(char ch)
  {
      // According to UTS#18 Unicode Regular Expressions (http://www.unicode.org/reports/tr18/)
      // RL 1.4 Simple Word Boundaries  The class of <word_character> includes all Alphabetic
      // values from the Unicode character database, from UnicodeData.txt [UData], plus the U+200C
      // ZERO WIDTH NON-JOINER and U+200D ZERO WIDTH JOINER.
      const char ZeroWidthNonJoiner = '\u200C', ZeroWidthJoiner = '\u200D';
      return IsWordChar(ch) || ch == ZeroWidthJoiner || ch == ZeroWidthNonJoiner;
  }
```

For branch prediction and common data, I would expect the ordering shown above to be more valuable than doing the _cheaper, but less likely_ character equality checks first...

**Comment by stephentoub** [comment]
> That last line shoudn't the  | ch == actually be  || ch == so we're doing a logical OR not a bitwise OR?

That is on purpose based on perf testing and the elimination of the branch on average being better.

**Comment by IDisposable** [comment]
Glad I asked! I've got the PR ready then... one other question. Would it be ugly to expose this constant in some way so the _Emitter_ generated code could just reference it? Is there precedence for that sort of construct?

**Comment by stephentoub** [comment]
> Would it be ugly to expose this constant in some way so the Emitter generated code could just reference it?

Yes.  We want the emitted code to be as readable as possible, which means it's better for it to read:
```C#
           const int WordCategories =
               1 << (int)UnicodeCategory.UppercaseLetter |
               1 << (int)UnicodeCategory.LowercaseLetter |
               1 << (int)UnicodeCategory.TitlecaseLetter |
               1 << (int)UnicodeCategory.ModifierLetter |
               1 << (int)UnicodeCategory.OtherLetter |
               1 << (int)UnicodeCategory.NonSpacingMark |
               1 << (int)UnicodeCategory.DecimalDigitNumber |
               1 << (int)UnicodeCategory.ConnectorPunctuation;
```
than to read:
```C#
           const int WordCategories = 0x4013F;
```

**Comment by EgorBo** [comment]
> than to read:

Reminds me this https://github.com/dotnet/runtime/blob/c3983a2ff0a18147af8d75dc3be4c22f76992bd7/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L90-L92

**Comment by IDisposable** [comment]
PR submitted, I did make a change in Emitter, but ONLY to change the name to `WordCategoriesMask` to better match the library code.

My follow-up question for the https://github.com/dotnet/runtime/pull/84003#issuecomment-1487474021 is more along the lines of should we expose a *public* constant from RegexCharClass that is _referenced_ in the Emitter's output code such that the emitted code looks like
```csharp
    "    return (uint)chDiv8 < (uint)ascii.Length ?",
    "        (ascii[chDiv8] & (1 << (ch & 0x7))) != 0 :",
    "        (RegexCharClass.WordCategoriesMask & (1 << (int)CharUnicodeInfo.GetUnicodeCategory(ch))) != 0;",
    "}",
});
```

**Comment by stephentoub** [comment]
> My follow-up question for the https://github.com/dotnet/runtime/pull/84003#issuecomment-1487474021 is more along the lines of should we expose a public constant from RegexCharClass that is referenced in the Emitter's output code such that the emitted code looks like

No, I don't believe there's enough value in that (and RegexCharClass is internal).

**Pull request merged**